src/io/DumpH5MDParallel.cpp

```diff
@@ -0,0 +1,479 @@
+
+    std::vector<hsize_t> dims = {1};
+    std::vector<int64_t> step = {0};
+    std::vector<double> time = {0};
+    std::string stepDataset = "/particles/atoms/" + forceDataset + "/step";
+    CHECK_HDF5(H5LTmake_dataset(fileId, stepDataset.c_str(), 1, dims.data(), typeToHDF5<int64_t>(),
+                                step.data()));
+    std::string timeDataset = "/particles/atoms/" + forceDataset + "/time";
+    CHECK_HDF5(H5LTmake_dataset(fileId, timeDataset.c_str(), 1, dims.data(), typeToHDF5<double>(),
+                                time.data()));
+    CHECK_HDF5(H5Gclose(group));
+}
+
+void DumpH5MDParallel::updateCache()
+{
+    boost::mpi::communicator world;
+    comm = world;
+    numProcesses = world.size();
+    rank = world.rank();
+
+    numLocalParticles = system_->storage->getNRealParticles();
+    MPI_Allreduce(reinterpret_cast<const void*>(&numLocalParticles),
+                  reinterpret_cast<void*>(&numTotalParticles), 1, MPI_INT64_T, MPI_SUM, comm);
+
+    MPI_Exscan(&numLocalParticles, &particleOffset, 1, MPI_INT64_T, MPI_SUM, comm);
+    if (rank == 0) particleOffset = 0;
+}
+
+void DumpH5MDParallel::dump()
+{
+    updateCache();
+
+    auto info = MPI_INFO_NULL;
+
+    auto plist = CHECK_HDF5(H5Pcreate(H5P_FILE_ACCESS));
+    CHECK_HDF5(H5Pset_fapl_mpio(plist, comm, info));
+
+    auto file_id = CHECK_HDF5(H5Fcreate(filename_.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, plist));
+
+    auto group1 =
+        CHECK_HDF5(H5Gcreate(file_id, "/particles", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT));
+    auto group2 =
+        CHECK_HDF5(H5Gcreate(file_id, "/particles/atoms", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT));
+
+    writeHeader(file_id);
+    writeBox(file_id);
+    if (dumpId) writeId(file_id);
+    if (dumpType) writeType(file_id);
+    if (dumpMass) writeMass(file_id);
+    if (dumpQ) writeQ(file_id);
+    if (dumpGhost) writeGhost(file_id);
+    if (dumpPosition) writePosition(file_id);
+    if (dumpVelocity) writeVelocity(file_id);
+    if (dumpForce) writeForce(file_id);
+
+    CHECK_HDF5(H5Gclose(group1));
+    CHECK_HDF5(H5Gclose(group2));
+
+    CHECK_HDF5(H5Fclose(file_id));
+}
+
+void DumpH5MDParallel::registerPython()
+{
+    using namespace espressopp::python;
+
+    class_<DumpH5MDParallel>("io_DumpH5MDParallel", init<shared_ptr<System>, std::string>())
+        .def_readwrite("author", &DumpH5MDParallel::author)
+        .def_readwrite("dumpId", &DumpH5MDParallel::dumpId)
+        .def_readwrite("dumpType", &DumpH5MDParallel::dumpType)
+        .def_readwrite("dumpMass", &DumpH5MDParallel::dumpMass)
+        .def_readwrite("dumpQ", &DumpH5MDParallel::dumpQ)
+        .def_readwrite("dumpGhost", &DumpH5MDParallel::dumpGhost)
+        .def_readwrite("dumpPosition", &DumpH5MDParallel::dumpPosition)
+        .def_readwrite("dumpVelocity", &DumpH5MDParallel::dumpVelocity)
+        .def_readwrite("dumpForce", &DumpH5MDParallel::dumpForce)
+        .def_readwrite("idDataset", &DumpH5MDParallel::idDataset)
```

Comment:
If this writer/restorer is for H5MD format then we shouldn't allow to modify the name of datasets as they are defined in the file format standard. See here: https://nongnu.org/h5md/h5md.html#particles-group
I would remove these properties or keep them as read-only.